pkg/kubernetes/watch/watcher.go

```diff
@@ -0,0 +1,318 @@
+	return w
+}
+
+// WithoutFieldSelector clears any field selector from this Watcher.
+// This is the default.
+func (w *Watcher[T]) WithoutFieldSelector() *Watcher[T] {
+	w.fieldSelector = ""
+	return w
+}
+
+// WithErrorCallback sets this Watcher's error callback. It's invoked every time
+// an error occurs and determines if the watch should continue or terminate:
+//   - The returned error is nil: continue watching
+//   - The returned error is not nil: terminate watch with that error
+//
+// If the error callback is not set or nil, the default behavior is to terminate.
+func (w *Watcher[T]) WithErrorCallback(callback func(error) error) *Watcher[T] {
+	w.errorCallback = callback
+	return w
+}
+
+// Until runs a watch until condition returns true. It will error out in case
+// the context gets canceled or the condition returns an error.
+func (w *Watcher[T]) Until(ctx context.Context, condition func(*T) (bool, error)) error {
+	listCondition := func(items []T) (bool, error) {
+		for i := range items {
+			ok, err := condition(&items[i])
+			if err != nil {
+				return false, err
+			}
+			if ok {
+				return true, nil
+			}
+		}
+
+		return false, nil
+	}
+
+	return retry(ctx, w.errorCallback, func(ctx context.Context) error {
+		return w.run(ctx, listCondition, condition)
+	})
+}
+
+func itemsFromList[L metav1.ListInterface, I any]() (func(L) []I, error) {
+	// List types from client-go don't provide any methods to get their items.
+	// Hence provide a way to get the items via reflection.
+
+	index, err := func() ([]int, error) {
+		var list L
+		var items []I
+		listType := reflect.TypeOf(list)
+		if listType.Kind() != reflect.Pointer {
+			return nil, fmt.Errorf("not a pointer type: %s", listType)
+		}
+		itemsType := reflect.TypeOf(items)
+		itemsField, ok := listType.Elem().FieldByName("Items")
+		if !ok || itemsField.Type != itemsType {
+			return nil, fmt.Errorf(`expected an "Items" field of type %s in %s`, itemsType, listType)
+		}
+		return itemsField.Index, nil
+	}()
+	if err != nil {
+		return nil, err
+	}
+
+	return func(l L) []I {
+		// The compatibility of the types has been checked above.
+		// This will not panic at runtime.
+		return reflect.ValueOf(l).Elem().FieldByIndex(index).Interface().([]I)
+	}, nil
+}
+
+type noRetry struct{ error }
+
+var errResourceTooOld = errors.New("resource too old")
+
+func (w *Watcher[T]) run(ctx context.Context, listCallback func([]T) (bool, error), watchCallback func(*T) (bool, error)) error {
+	var initialResourceVersion string
+	{
+		resourceVersion, items, err := w.List(ctx, metav1.ListOptions{
+			FieldSelector:  w.fieldSelector,
+			TimeoutSeconds: pointer.Int64(30),
+		})
+		if err != nil {
+			return err
+		}
+		if ok, err := listCallback(items); err != nil {
+			return noRetry{err}
+		} else if ok {
+			return nil
+		}
+
+		initialResourceVersion = resourceVersion
+	}
+
+	watcher, err := watchtools.NewRetryWatcher(initialResourceVersion, &cache.ListWatch{
```

Comment:
This was used in the original autopilot code, but had to be abandoned in favor for the polling that you're removing due to extremely hard-to-reproduce scenarios where the watcher wouldn't trigger (despite data being there).

It is the better direction, so I'm hopeful that this diff (+ time) has more success than I did (PEBKAC).

Reply:
Sounds a little bit concerning. I thought it would be good to use the watcher everywhere, so it gets "battle tested", sorta. Are there some references left to the original watching code, maybe in the git history? I would love to have a look at it. Don't want to shoot ourselves into the foot with this change.

Do you have some scenarios at hand that triggered the bad behavior?